nan.h

```diff
@@ -1713,6 +1733,84 @@ template<class T>
 // class definition.
 typedef AsyncProgressWorkerBase<char> AsyncProgressWorker;
 
+template<class T>
+/* abstract */
+class AsyncProgressQueueWorker : public AsyncBareProgressWorker<T> {
+ public:
+  explicit AsyncProgressQueueWorker(Callback *callback_)
+      : AsyncBareProgressWorker<T>(callback_) {
+    uv_mutex_init(&async_lock);
+  }
+
+  virtual ~AsyncProgressQueueWorker() {
+    uv_mutex_lock(&async_lock);
+
+    while (!asyncdata_.empty()) {
+      std::pair<T*, size_t> *datapair = asyncdata_.front();
+      T *data = datapair->first;
+
+      asyncdata_.pop();
+
+      delete[] data;
+      delete datapair;
+    }
+
+    uv_mutex_unlock(&async_lock);
+    uv_mutex_destroy(&async_lock);
+  }
+
+  void WorkComplete() {
+    WorkProgress();
+    AsyncWorker::WorkComplete();
+  }
+
+  void WorkProgress() {
+    uv_mutex_lock(&async_lock);
+
+    while (!asyncdata_.empty()) {
+      std::pair<T*, size_t> *datapair = asyncdata_.front();
+      asyncdata_.pop();
+      uv_mutex_unlock(&async_lock);
+
+      T *data = datapair->first;
+      size_t size = datapair->second;
+
+      // Don't send progress events after we've already completed.
+      if (this->callback) {
+          this->HandleProgressCallback(data, size);
+      }
+
+      delete[] data;
+      delete datapair;
+
+      uv_mutex_lock(&async_lock);
+    }
+
+    uv_mutex_unlock(&async_lock);
+  }
+
+ private:
+  void SendProgress_(const T *data, size_t count) {
+    T *new_data = new T[count];
+    {
+      T *it = new_data;
+      std::copy(data, data + count, it);
+    }
+
+    std::pair<T*, size_t> *datapair =
+      new std::pair<T*, size_t>(new_data, count);
+
+    uv_mutex_lock(&async_lock);
+    asyncdata_.push(datapair);
+    uv_mutex_unlock(&async_lock);
+
+    uv_async_send(this->async);
+  }
+
+  uv_mutex_t async_lock;
+  std::queue<std::pair<T*, size_t>*> asyncdata_;
```

Comment:
This doesn't need to be a pointer to a `std::pair<...>`, does it?

Reply:
Is something wrong with a pointer to a std::pair<> ?  I wanted this new class template to be a drop-in replacement for `AsyncProgressWorkerBase`, which means it must pass both a `T*` and a `size_t` , I thought this was the natural way to do it.  Would you prefer a `std::pair<T*, size_t>` instead of a pointer?

Reply:
Yes, that's what I mean, the extra level of indirection is not needed.  Sorry, I should have been more explicit.

Reply:
no problem. done in 02ccbd5c279df612bd552765fb5da1fe1da1d234

Reply:
no problem.  done in 8fb5ba91d05e2631523d97365c7c8333ad79e25b